### PR TITLE
RAINCATCH-460 - Mongoose-store initial implementation and cleanup

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Provide Mongoose backed storage along with schemas and validation. 
 
ping @darachcawley @col1985

### CODE REVIEW

File: lib/listen.js [hunk 1]
```diff
@@ -3,47 +3,54 @@ var Topics = require('fh-wfm-mediator/lib/topics');
...
+        }).catch(function(error) {
+          console.log(error);
+        });
       })
+      .on('list', function(filter) {
+        self.list(filter).then(function(object) {
+          self.mediator.publish(self.topics.getTopic('list', 'done'), object);
+        }).catch(function(error) {
+          console.log(error);
+        });
```
Comment (paolobueno) [inline lib/listen.js]: If you `catch` errors and don't `return` the promise it'll just swallow the error and won't allow for `catch` calls in client code to handle the errors.

File: lib/mongoose-store.js [hunk 2]
```diff
@@ -1,94 +1,64 @@
...
+    console.log("Initialization data is not array.");
+    return Promise.resolve();
+  }
+  var self = this;
+  var promises = [];
+  _.each(data, function(entry) {
+    var record = new self.model(entry);
+    promises.push(record.save());
+  });
+  return Promise.all(promises);
```
Comment (paolobueno) [inline lib/mongoose-store.js]: You can use `Promise.map` from bluebird instead of `_.each() + push`

**Review by paolobueno — DISMISSED** [review]
I think the current swallowing of exceptions is a blocker for merging.

**Comment by paolobueno** [comment]
I think we should try to push data schemas to the individual modules using a generic schema-validation tool instead of providing that functionality only through the mongodb store.

File: lib/mongoose-store.js [hunk 3]
```diff
@@ -1,94 +1,64 @@
...
-  return record.save(); //.then(_.property('fields'));
+  var record = new this.model(object);
+  return record.save();
 };
 
 Store.prototype.findById = function(id) {
-  console.log('do findById--- ', this.datasetId + ' :id: ' + id);
   // use 'list' since 'read' can only filter by guid === mongo's ObjectId
-  return this.model.findOne({id: id}).then(extractListFields);
+  return this.model.findOne({id: id}).lean();
```
Comment (paolobueno) [inline lib/mongoose-store.js]: I notice we're using `lean()` everywhere.
This strips all extra mongoose functionality. I think this is related to providing the same API as the `simple-store` so a :+1: on this, however doesn't this mean we could be achieving better performance through a lower-level api like the `mongodb` module itself?
Assuming we push schema validation to the individual modules, of course.

**Comment by wtrocki** [comment]
@paolobueno There is little bit context behind this changes. We currently working on this PR to resolve listener part to sent promises. Fully agree that we should move schemas to modules and we planning to do it as part of different issue, but this is not target of this PR. What I'm trying to do is to clean up current implementation and resolve all blockers.

TL;DR - this PR is just scratching surface.

File: lib/mongoose-store.js (hunk 3 above)
Comment (wtrocki) [reply lib/mongoose-store.js]: That's something we need to rethink as mongose related fields would introduce problems with sync (conflicts). For now we assume that we are only interested in plain objects, which is way faster than wrapping them into the models. We not going to use models in any way as they are often passed unchanged with little to no changes that needed to be passed back to db. I think that lean is the best option to use for the moment considering limitations and requirements.

File: lib/listen.js (hunk 1 above)
Comment (wtrocki) [reply lib/listen.js]: Moved this back to promises but still working on other issues

**Comment by wtrocki** [comment]
Reviewed. Problems resolved directly in comments
